tsconfig.json

```diff
@@ -1,7 +1,6 @@
 {
   "extends": "@tsconfig/node18/tsconfig.json",
   "compilerOptions": {
-    "module": "commonjs",
```

Comment:
For context, as the failing build log will eventually expire, this is the error message:

```
Error: tsconfig.json(4,15): error TS5110: Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'.
```

This error is new to TypeScript 5.2.2, and comes from this PR: https://github.com/microsoft/TypeScript/pull/54567 .

The relevant bits of configuration from `@tsconfig/node18/tsconfig.json`:

```json
{
  "compilerOptions": {
    "module": "Node16",
    "moduleResolution": "node16"
  }
}
```

This was also recently changed: in https://github.com/tsconfig/bases/pull/197, which we brought in via https://github.com/PhilanthropyDataCommons/service/pull/453 .

I do not yet fully understand the consequences of opting in to `node16` for `module` and `moduleResolution` (see also the [docs for `module` in `tsconfig.json`](https://www.typescriptlang.org/tsconfig#module)). My concern is that we'd be opening the door to ESM, which I would very much like to avoid.

I'll keep researching, but wanted to document how we got to this question in the first place. Despite being a single line change, this is a complicated situation!

Reply:
@jasonaowen thanks for doing this research -- once the dust settles I'll make sure to update the commit message with the information (and add you as a co-author)

Reply:
Update here -- reading through https://www.typescriptlang.org/docs/handbook/esm-node.html

It looks like Node16 means typescript *SUPPORTS* ESM, not that it must be ESM.  Specifically, Node16 means that if `type: 'module'` is defined in `package.json` then typescript will start to flag ESM errors.

This is why the PR does not currently error (we don't specify `type` in package.json so it is using the default, which is `commonjs`).

If we add `"type": "module"` to `package.json` then running build results in tons of ESM errors:

<img width="983" alt="image" src="https://github.com/PhilanthropyDataCommons/service/assets/208884/3981b0b3-199c-4cb1-8eba-736752cec8af">

This is corroborated with the [TS documentation](https://www.typescriptlang.org/tsconfig#node16nodenext-nightly-builds) Which notes:

```quote
Available from 4.7+, the node16 and nodenext modes integrate with Node’s [native ECMAScript Module support](https://nodejs.org/api/esm.html). The emitted JavaScript uses either CommonJS or ES2020 output depending on the file extension and the value of the type setting in the nearest package.json. Module resolution also works differently. You can learn more in the [handbook](https://www.typescriptlang.org/docs/handbook/esm-node.html).
```

Reply:
Thank you, @slifty! I think that matches my initial research.

- does this allow us to use ESM-only libraries? what happens if we do?
- are there any drawbacks to setting `"type": "commonjs"`? if not, I really like that idea!

Reply:
@jasonaowen 

> does this allow us to use ESM-only libraries? what happens if we do?

This is a great question!  I'll test it out (my theory is no since it has to be able to compile to common js. stay tuned to find out)

> are there any drawbacks to setting "type": "commonjs"? if not, I really like that idea!

My understanding of the documentation of `type` is that omission is the same as explicitly setting to `commonjs` so let's do it!

Reply:
The behavior on this branch appears to the same as `main` regarding esm-only modules still error because we're writing cjs.  I tried installing `node-fetch` on both and in both cases it triggered the same output when running `build` both for when type was `commonjs` and when type was `module`.

This makes sense to me because how imports are processed [is dictated by the `moduleResolution` setting](https://www.typescriptlang.org/docs/handbook/module-resolution.html) (as opposed to the `module` setting).  The `moduleResolution` is the same on this branch and on `main (it is `Node16`).

Reply:
Thank you for following up on this, @slifty!

> This makes sense to me because the way imports are processed [is dictated by the `moduleResolution` setting](https://www.typescriptlang.org/docs/handbook/module-resolution.html) (as opposed to the `module` setting). The `moduleResolution` is the same on this branch and on `main` (it is `Node16`).

Note that that changed recently, in #412 - previously the `moduleResolution` was `node`, which the [TypeScript docs](https://www.typescriptlang.org/tsconfig#moduleResolution) describe as

> 'node10' (previously called 'node') for Node.js versions older than v10, which only support CommonJS require. You probably won’t need to use node10 in modern code.

I ran through the manual test you described with commit 86ea0c9 (the parent commit of #412), and the new behavior is better: previously, it would successfully build, but then at runtime say

```
Error [ERR_REQUIRE_ESM]: require() of ES Module node_modules/node-fetch/src/index.js from dist/index.js not supported.
Instead change the require of node_modules/node-fetch/src/index.js in dist/index.js to a dynamic import() which is available in all CommonJS modules.
    at Object.<anonymous> (dist/index.js:8:38) {
  code: 'ERR_REQUIRE_ESM'
}
```

So, +1 to this change, and thanks again @slifty for digging in!